nntrainer/layers/embedding.cpp

```diff
@@ -133,10 +135,38 @@ void EmbeddingLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 void EmbeddingLayer::calcDerivative() {
-  // NYI
+  return; // intended
 }
+
 void EmbeddingLayer::calcGradient() {
-  // NYI
+  Tensor &djdw =
+    weightAt(static_cast<int>(EmbeddingParams::weight)).getGradientRef();
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  djdw.setZero();
+
+  // TODO:
+  // This is to calculate gradient with current implementation of optimizer.
+  // In order to accelerate, we need to better way like using index to weight.
+
+  for (unsigned int b = 0; b < input_.batch(); ++b) {
+    float *in_data = input_.getAddress(b * input_.getDim().getFeatureLen());
+
+    for (unsigned int i = 0; i < in_length; ++i) {
+      if (in_data[i] - 1 < 0)
+        continue;
+
+      float *djdw_data =
+        djdw.getAddress(static_cast<uint>(in_data[i] - 1) * out_dim);
+      float *grad_data = derivative_.getAddress(
+        b * derivative_.getDim().getFeatureLen() + i * out_dim);
+
+      for (unsigned int j = 0; j < out_dim; ++j) {
```

Comment:
please use memset here instead of a loop.

Reply:
It is not assign. It addition.

Reply:
then please use something like `std::transform (djdw.begin(), djdw.end(), grad_data.begin(), djdw.begin(), std::plus<float>());`

Reply:
does it faster than this? if it is, let me know.

Reply:
yes, it is faster with `std::transform()` as it is easier for compile to optimize these, and from c++17, allows parallel computation with transform.

Reply:
done